go/ai-proxy/proxy/ollama/ollama.go

```diff
@@ -0,0 +1,319 @@
+				},
+			},
+		}
+
+		payload, err := json.Marshal(completionChunk)
+		if err != nil {
+			return err
+		}
+
+		_, _ = fmt.Fprintf(w, "data: %s\n\n", payload)
+		flusher.Flush()
+		return nil
+	}
+
+	err = o.ollamaClient.Chat(context.Background(), input, respFunc)
+	if err != nil {
+		klog.ErrorS(err, "failed to stream ollama response")
+		return
+	}
+
+	// Send OpenAI's `[DONE]` event to signal end of stream
+	_, _ = fmt.Fprintf(w, "data: [DONE]\n\n")
+	flusher.Flush()
+}
+
+func (o *OllamaProxy) handleNonStreamingOllama(
+	w http.ResponseWriter,
+	req *openai.ChatCompletionRequest,
+) {
+	input, err := o.convertOpenAIToOllamaChatRequest(req)
+	if err != nil {
+		klog.ErrorS(err, "failed to convert ollama request")
+		return
+	}
+
+	var ollamaResponse ollamaapi.ChatResponse
+	respFunc := func(output ollamaapi.ChatResponse) error {
+		ollamaResponse = output
+		return nil
+	}
+
+	err = o.ollamaClient.Chat(context.Background(), input, respFunc)
+	if err != nil {
+		klog.ErrorS(err, "failed to get ollama response")
+		return
+	}
+
+	response := openai.ChatCompletion{
+		Id:      uuid.NewString(),
+		Object:  ChatCompletionObject,
+		Created: time.Now().Unix(),
+		Model:   req.Model,
+		Choices: []openai.Choice{
+			{
+				Index: 0,
+				Message: func() openai.Message {
+					message := openai.Message{
+						Role:    "assistant",
+						Content: ollamaResponse.Message.Content,
+					}
+
+					if len(ollamaResponse.Message.ToolCalls) > 0 {
+						var toolCalls []openai.ToolCall
+						for _, tc := range ollamaResponse.Message.ToolCalls {
+							_, index, id, exists := o.toolRegistry.Get(tc.Function.Name)
+							if !exists {
+								klog.ErrorS(fmt.Errorf("tool not found"), "tool does not exist in registry", "tool", tc.Function.Name)
+								continue
+							}
+
+							argBytes, err := json.Marshal(tc.Function.Arguments)
+							if err != nil {
+								klog.ErrorS(err, "failed to marshal tool call arguments")
+								continue
+							}
+
+							toolCalls = append(toolCalls, openai.ToolCall{
+								ID:    id,
+								Index: index,
+								Function: struct {
+									Name      string `json:"name"`
+									Arguments string `json:"arguments"`
+								}{
+									Name:      tc.Function.Name,
+									Arguments: string(argBytes),
+								},
+							})
+						}
+						message.ToolCalls = toolCalls
+					}
+
+					return message
+				}(),
+			},
+		},
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	if err := json.NewEncoder(w).Encode(response); err != nil {
+		klog.Errorf("Error encoding response: %v", err)
+		return
+	}
+
+}
+
+func (o *OllamaProxy) convertOpenAIToOllamaChatRequest(req *openai.ChatCompletionRequest) (*ollamaapi.ChatRequest, error) {
+	var messages []ollamaapi.Message
+
+	for _, tool := range req.Tools {
+		otc := ollamaapi.ToolCall{
+			Function: ollamaapi.ToolCallFunction{
+				Name: tool.Function.Name,
+			},
+		}
+		if _, _, _, exists := o.toolRegistry.Get(otc.Function.Name); !exists {
```

Comment:
I think here you can just call `o.toolRegistry.Register` since it does the existence check internally